File: Xamarin/Xamarin.Native/Xamarin.Native.iOS/ViewControllers/NetworkTestController.cs
using CoreFoundation;
using Foundation;
using System;
using System.Diagnostics;
using System.Drawing;
using UIKit;
using Xamarin.Services;

namespace Xamarin.Native.iOS.ViewControllers
{
    public partial class NetworkTestController : UIViewController
    {
        Stopwatch stopwatch;
        NetworkTestService networkService;

        public NetworkTestController(IntPtr handle) : base(handle)
        {
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            AddressField.Text = "http://cdn.superbwallpapers.com/wallpapers/meme/doge-pattern-27481-2880x1800.jpg";
            networkService = new NetworkTestService();
            networkService.ImageDownloadCompleted += ImageDownloadCompleted;
        }

        partial void StartDownloading(UIButton sender)
        {
            stopwatch = new Stopwatch();
            stopwatch.Start();
            RefreshUI(true);

            DispatchQueue.DefaultGlobalQueue.DispatchAsync(() =>
            {
                networkService.DownloadImage(AddressField.Text);
            });
        }

        private void ImageDownloadCompleted(UIImage image)
        {
            stopwatch.Stop();
            DispatchQueue.MainQueue.DispatchAsync(() =>
            {
                DownloadedImage.Image = image;
                TimeLabel.Text = stopwatch.GetDurationInSeconds();
                RefreshUI(false);
            });
        }

        private void RefreshUI(bool isDownloading)
        {
            StartButton.Hidden = isDownloading;
            AddressField.Enabled = isDownloading;
            ActivityIndicator.Hidden = !isDownloading;
            ActivityIndicator.StartAnimating();
        }
    }
}

File: Native.iOS/Native.iOS/ViewControllers/NetworkTestController.m
//
//  NetworkTestController.m
//  Native.iOS
//
//  Created by Pawel Grzmil on 17/01/16.
//  Copyright © 2016 Pawel Grzmil. All rights reserved.
//

#import "NetworkTestController.h"
#import "NetworkTestService.h"
#import "Stopwatch.h"

@interface NetworkTestController()<NetworkTestServiceDelegate>

@property (weak, nonatomic) IBOutlet UILabel *timeLabel;
@property (weak, nonatomic) IBOutlet UIButton *startButton;
@property (weak, nonatomic) IBOutlet UITextField *addressField;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (weak, nonatomic) IBOutlet UIImageView *downloadedImage;

@property (strong, nonatomic) Stopwatch *stopwatch;
@property (strong, nonatomic) NetworkTestService *networkService;

@end
@implementation NetworkTestController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Test obsługi sieci";
    self.addressField.text = @"http://cdn.superbwallpapers.com/wallpapers/meme/doge-pattern-27481-2880x1800.jpg";
    
    self.stopwatch = [Stopwatch new];
    
    self.networkService = [NetworkTestService new];
    self.networkService.delegate = self;
}

- (IBAction)startDownloading:(id)sender {
    self.downloadedImage.image = nil;
    [self refreshUI:true];
    [self.stopwatch start];
    
    dispatch_async( dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self.networkService downloadImage:self.addressField.text];
    });
}

- (void)imageDownloadCompleted:(UIImage *)image{
    [self.stopwatch stop];
    dispatch_async( dispatch_get_main_queue(), ^{
        self.downloadedImage.image = image;
        self.timeLabel.text = [self.stopwatch getDurationInSeconds];
        [self refreshUI:false];
    });
}

- (void)refreshUI:(BOOL)isDownloading{
    self.activityIndicator.hidden = !isDownloading;
    self.startButton.hidden = isDownloading;
    self.addressField.userInteractionEnabled = !isDownloading;
}

@end
